python/ray/ml/preprocessors/encoder.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Dict
+
+import pandas as pd
+
+from ray.data import Dataset
+from ray.data.aggregate import UniqueValues
+from ray.ml.preprocessor import Preprocessor
+
+
+class OrdinalEncoder(Preprocessor):
+    """Encode values within columns as ordered integer values.
+
+    Currently, order within a column is based on the values from the fitted
+    dataset in sorted order.
+
+    Transforming values not included in the fitted dataset will be encoded as ``None``.
+
+    Args:
+        columns: The columns that will individually be encoded.
+    """
+
+    def __init__(self, columns: List[str]):
+        # TODO: allow user to specify order of values within each column.
+        super().__init__()
+        self.columns = columns
+
+    def _fit(self, dataset: Dataset) -> Preprocessor:
+        aggregates = [UniqueValues(col) for col in self.columns]
+        stats = dataset.aggregate(*aggregates)
+        self.stats = {row: _sorted_value_indices(stats[row]) for row in stats}
+        return self
+
+    def _transform_pandas(self, df: pd.DataFrame):
+        def column_ordinal_encoder(s: pd.Series):
+            s_values = self.stats[f"unique_values({s.name})"]
+            return s.map(s_values)
+
+        df.loc[:, self.columns] = df.loc[:, self.columns].transform(
+            column_ordinal_encoder
+        )
+        return df
+
+    def __repr__(self):
+        return f"<Encoder columns={self.columns} stats={self.stats}>"
+
+
+class OneHotEncoder(Preprocessor):
+    """Encode columns as new columns using one-hot encoding.
+
+    The transformed dataset will have a new column in the form ``{column}_{value}``
+    for each of the values from the fitted dataset. The value of a column will
+    be set to 1 if the value matches, otherwise 0.
+
+    Transforming values not included in the fitted dataset will result in all
+    of the encoded column values to be 0.
+
+    Args:
+        columns: The columns that will individually be encoded.
+    """
+
+    def __init__(self, columns: List[str]):
+        super().__init__()
+        self.columns = columns
+
+    def _fit(self, dataset: Dataset) -> Preprocessor:
+        aggregates = [UniqueValues(col) for col in self.columns]
+        stats = dataset.aggregate(*aggregates)
+        self.stats = {row: _sorted_value_indices(stats[row]) for row in stats}
+        return self
+
+    def _transform_pandas(self, df: pd.DataFrame):
+        for column in self.columns:
+            column_values = self.stats[f"unique_values({column})"]
+            for column_value in column_values:
+                df[f"{column}_{column_value}"] = df[column].map(
+                    lambda x: int(x == column_value)
+                )
+
+        # Drop original unencoded columns.
+        df = df.drop(columns=self.columns)
```

Comment:
Nit: Might this cause the intermediate dataframe to get pretty large if many columns are being one-hot encoded? Not sure how big of an issue this would be.

Reply:
We could probably drop first and then add new columns

Reply:
Updated to compute new values => drop old values => add new values.

Reply:
This is still materializing all of the one-hot encoded columns and the old columns at the same time, just with the new one-hot encoded columns being in a dictionary instead of the DataFrame.

I'd vote for either keeping the one-hot encoding as you had it before, with columns added directly into the DataFrame, until it becomes a problem:
```python
    def _transform_pandas(self, df: pd.DataFrame):
        _validate_df(df, *self.columns)
        # Compute new one-hot encoded columns
        for column in self.columns:
            column_values = self.stats_[f"unique_values({column})"]
            for column_value in column_values:
                df[f"{column}_{column_value}"] = (
                    df[column] == column_value
                ).astype(int)
        # Drop original unencoded columns.
        df = df.drop(columns=self.columns)
        return df
```
Or dropping columns progressively and in-place as the corresponding one-hot encoded columns are added:
```python
    def _transform_pandas(self, df: pd.DataFrame):
        _validate_df(df, *self.columns)
        # Compute new one-hot encoded columns
        for column in self.columns:
            column_values = self.stats_[f"unique_values({column})"]
            for column_value in column_values:
                df[f"{column}_{column_value}"] = (
                    df[column] == column_value
                ).astype(int)
            df.drop(columns=[column], inplace=True)
        return df
```
I'd vote for the former until this becomes an issue.

Reply:
Ack changing back for now.